build-tools/packages/version-tools/src/schemes.ts

```diff
@@ -0,0 +1,48 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import * as semver from "semver";
+import { VersionScheme } from "@fluidframework/build-tools";
+import { isInternalVersionScheme } from "./internalVersionScheme";
+
+/**
+ * Given a version or a range string, determines what version scheme the string is using.
+ * @param rangeOrVersion - a version or range string.
+ * @returns The version scheme that the string is in.
+ */
+export function detectVersionScheme(rangeOrVersion: string): VersionScheme {
+    // First check if the string is a valid internal version
+    if (isInternalVersionScheme(rangeOrVersion)) {
+        return "internal";
+    }
+
+    if (semver.valid(rangeOrVersion) !== null) {
+        // Must be a version string
+        // If the major is less than 0 assume it's a virtualPatch version
+        if (semver.major(rangeOrVersion) === 0) {
```

Comment:
would `major` throw if the `rangeOrVersion` string was invalid (ex "badversion") ?

Reply:
According to the docs, nothing is thrown. But it should be protected by the call to semver.valid -- at the point major is called, we've verified that semver thinks it's a version string.